infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuReplicationDiscoveryService.java

```diff
@@ -517,18 +537,35 @@ public void updateTopology(LogReplicationClusterInfo.TopologyConfigurationMsg to
      * msg needs to send out.
      */
     @Override
-    public void prepareSiteRoleChange() {
-        replicationManager.prepareSiteRoleChange();
+    public void prepareClusterRoleChange() {
+        if (localClusterDescriptor.getRole() == ClusterRole.ACTIVE && replicationManager != null) {
+            replicationManager.prepareClusterRoleChange();
+        } else {
+            log.warn("Illegal prepareSiteRoleChange when cluster{} with role {}",
+                    localClusterDescriptor.getClusterId(), localClusterDescriptor.getRole());
+        }
     }
 
     /**
-     * Query all replicated stream log tails and calculate the number of messages to be sent.
-     * If the max tail has changed, return 0%.
+     * Query the current all replication stream log tail and calculate the number of messages to be sent.
+     * If the max tail has changed, give 0%.
      */
     @Override
     public int queryReplicationStatus() {
         // TODO (maxi): address Nan's comments
-        return replicationManager.queryReplicationStatus();
+        if (localClusterDescriptor.getRole() == ClusterRole.ACTIVE) {
+            // If not leader, replication manager might be null
+            if (replicationManager != null) {
```

Comment:
I think we should only reply if we are the leader. Cause from their point of view most likely we'll tell them to query all nodes in the cluster and they'll pick the max of these 3 values, otherwise, they won't know who is the leader right now as the lock is something internal.

Reply:
Added check and logs